Restart pods on changing imported BOSH properties

mook-as opened the pull request:
This implements pods (on changing their exported properties) restarting pod which are known to have a dependency on the changed properties.

https://github.com/cloudfoundry-incubator/fissile/pull/470 implements tracking this dependency information in the images generated by fissile; this PR makes configgin read those changes and sets the digests on the statefulsets which import those properties.  This also makes configgin set the properties on import time, such that if we hit a race where a pod manages to import the properties before the exporting pod changes the statefulset we will not be restarted.

This also includes a changeset to silence much of the rubocop lint warnings (excluding the metrics categories that will require refactoring).

mook-as committed:
[WIP] Restart pods on imported property change

Currently just have code to mark property hashes on export and copy the hash to
the pods on import.  The hashing code is broken (it always hashes "null" instead
of the correct value).

Once this is done, we need to (from the pod that does the exporting) set the
same hashes on statefulsets which import those properties, under the
.spec.metadata.annotations key, such that kube will restart pods as appropriate.

mook-as committed:
Actually patch the StatefulSet as necessary on prop change

This actually implements modifying the dependent StatefulSets when the
exported properties change.  Kube is expected to restart pods as
appropriate.

This commit breaks all the tests; they will be fixed before PR.

mook-as committed:
Fix existing tests to work with changes

mook-as committed:
Use explicit kwargs for Configgin.new

mook-as committed:
Spec: testing for patching statefulset on pod property change

mook-as committed:
Rubocop fixes

This still leaves various code metrics-related linting issues.

mook-as commented:
Dropped the bundler bump, because CI seems to be having issues with it.

f0rmiga commented on lib/configgin.rb:
```diff
@@ -41,19 +46,35 @@ def generate_jobs(job_configs, templates)
   end
 
-  def set_job_metadata(jobs)
+  # Set the exported properties and their digests, and return the digests
```
[nit] missing punctuation.

mook-as replied:
Fixed in da315a15408eec9730f51d8d295c513d19f22eb2

f0rmiga commented on lib/configgin.rb:
```diff
@@ -41,19 +46,35 @@ def generate_jobs(job_configs, templates)
-        ENV['HOSTNAME'],
-        { metadata: { annotations: { :"skiff-exported-properties-#{name}" => job.exported_properties.to_json } } },
+        @self_name,
+        { metadata: {
```
This formatting is confusing.

mook-as replied:
Fixed in da315a15408eec9730f51d8d295c513d19f22eb2

f0rmiga commented on lib/configgin.rb:
```diff
@@ -66,6 +87,51 @@ def render_job_templates(jobs, job_configs)
     end
   end
 
+  # Some pods might have depended onthe properties exported by this pod; locate
```
onthe -> on the.

mook-as replied:
Fixed in da315a15408eec9730f51d8d295c513d19f22eb2

f0rmiga commented on lib/configgin.rb:
```diff
@@ -66,6 +87,51 @@ def render_job_templates(jobs, job_configs)
+          end
+          # rubocop:enable Layout/RescueEnsureAlignment
+          if response['reason'] == 'NotFound'
+            # The StatefulSet can be missing if we're configured to not have an optional instance group
```
[nit] Punctuation.

mook-as replied:
Fixed in da315a15408eec9730f51d8d295c513d19f22eb2

f0rmiga commented on lib/kube_link_generator.rb:
```diff
@@ -66,17 +69,26 @@ def get_pods_for_role(role_name, job, options = {})
+  def get_exported_properties(role_name, pod, job_name)
+    if pod.metadata.annotations["skiff-exported-properties-#{job_name}"]
+      if pod.metadata.annotations["skiff-exported-digest-#{job_name}"]
+        # Copy the digest over, so that if the source role changes we can be restarted
```
[nit] Punctuation.

mook-as replied:
Fixed in da315a15408eec9730f51d8d295c513d19f22eb2

f0rmiga commented on lib/configgin.rb:
```diff
@@ -41,19 +46,35 @@ def generate_jobs(job_configs, templates)
 
-  def set_job_metadata(jobs)
+  # Set the exported properties and their digests, and return the digests
+  def patch_job_metadata(jobs)
```
How about a unit test for this function?

mook-as replied:
Added in 4b955ff2103a637fd1352e2fdfec06a46fcff390

f0rmiga commented on lib/configgin.rb:
```diff
@@ -66,6 +87,51 @@ def render_job_templates(jobs, job_configs)
 
+  # Some pods might have depended onthe properties exported by this pod; locate
+  # them and cause them to restart as appropriate.
+  def restart_affected_pods(job_configs, job_digests)
```
There is quite a bit of logic in this one. How about a unit test for it?

mook-as replied:
Refactored and added in 26897e5e22937ebd5e0e1cb5ca5f15c4fffad2a3

f0rmiga reviewed (changes requested):
Overall looks very good. Just a couple unit tests requested and a few nits.

mook-as committed:
Address (trivial) review comments

https://github.com/cloudfoundry-incubator/configgin/pull/97#pullrequestreview-222982330

Follow-up commits will add the requested tests.

mook-as committed:
spec: test Configgin#patch_job_metadata

mook-as committed:
spec: test Configgin#restart_affected_pods

This includes refactoring the method to pull out calculating the
expected annotations, so that they can be tested separately.

mook-as reviewed (commented):
Comments should be addressed (I think?)
